dont-wreck-my-house/src/main/java/learn/lodging/domain/GuestService.java

```diff
@@ -0,0 +1,138 @@
+        return repository.findAll().stream()
+                .filter(i -> i.getLastName().startsWith(prefix))
+                .collect(Collectors.toList());
+    }
+
+    public Result<Guest> add(Guest guest) throws DataException {
+        Result<Guest> result = validate(guest);
+        if (!result.isSuccess()){
+            return result;
+        }
+        guest.setPhoneNum(formatPhoneNumber(guest.getPhoneNum()));
+        result.setPayload(repository.add(guest));
+        return result;
+    }
+
+    public Result<Guest> update(Guest guest) throws DataException {
+        Result<Guest> result = validate(guest);
+        if (!result.isSuccess()){
+            return result;
+        }
+        List<Guest> guests = repository.findAll();
+        Guest existing = null;
+        for (Guest g: guests){
+            if (g.getId() == guest.getId()){
+                existing = g;
+            }
+        }
+        if (existing == null){
+            result.addErrorMessage("Guest not found.");
+            return result;
+        }
+        guest.setPhoneNum(formatPhoneNumber(guest.getPhoneNum()));
+        boolean success = repository.update(guest);
+        if (!success){
+            result.addErrorMessage("Guest not found.");
+        }
+        return result;
+    }
+
+    public Result<Guest> delete(Guest guest) throws DataException {
+        Result<Guest> result = new Result<>();
+        boolean success = repository.delete(guest);
+        if (!success){
+            result.addErrorMessage("Guest not found.");
+        }
+        return result;
+    }
+
+    private Result<Guest> validate(Guest guest){
+        Result<Guest> result = validateNulls(guest);
+        if (!result.isSuccess()){
+            return result;
+        }
+        validateInputs(guest,result);
+        return result;
+    }
+
+    private Result<Guest> validateNulls(Guest guest){
+        Result<Guest> result = new Result<>();
+        if (guest == null){
+            result.addErrorMessage("Nothing to save.");
+            return result;
+        }
+        if (guest.getFirstName() == null){
+            result.addErrorMessage("First Name is required.");
+        }
+        if (guest.getLastName() == null){
+            result.addErrorMessage("Last Name is required.");
+        }
+        if (guest.getEmail() == null){
+            result.addErrorMessage("Email Address is required.");
+        }
+        if (guest.getPhoneNum() == null){
+            result.addErrorMessage("Phone # is required.");
+        }
+        if (guest.getState() == null){
+            result.addErrorMessage("State is required.");
+        }
+        return result;
+    }
+
+    private void validateInputs(Guest guest, Result<Guest> result){
+
+        if (!(guest.getEmail().contains("@") && guest.getEmail().contains("."))){
+            result.addErrorMessage("Email must be a valid email address (containing '@' and '.').");
+        }
```

Comment:
Nice consideration. Check out regular expressions for some very thorough validations (research topic)